diff --git a/ttyio.c b/ttyio.c
index a1a13b1..0ba26c2 100644
--- a/ttyio.c
+++ b/ttyio.c
@@ -125,7 +125,7 @@
 #    include <starlet.h>
 #    include <ssdef.h>
 #  else /* !VMS */
-#    ifdef HAVE_TERMIOS_H
+#  if defined ( HAVE_TERMIOS_H ) || defined ( __MVS__ )
 #      include <termios.h>
 #      define sgttyb termios
 #      define sg_flags c_lflag
@@ -152,6 +152,7 @@
 #          if (!defined(MINIX) && !defined(GOT_IOCTL_H))
 #            include <sys/ioctl.h>
 #          endif
+#        ifndef __MVS__
 #          include <sgtty.h>
 #          define GTTY gtty
 #          define STTY stty
@@ -165,6 +166,7 @@
              int stty OF((int, struct sgttyb *));
               */
 #          endif
+#         endif
 #        endif /* !CMS_MVS */
 #      endif /* ?USE_SYSV_TERMIO */
 #    endif /* ?HAVE_TERMIOS_H */
diff --git a/unix/unix.c b/unix/unix.c
index efa97fc..a20d085 100644
--- a/unix/unix.c
+++ b/unix/unix.c
@@ -62,7 +62,7 @@
 #      include <ndir.h>
 #    endif
 #  else /* !SYSV */
-#    ifndef NO_SYSDIR
+#  if ! defined ( NO_SYSDIR ) && ! defined ( __MVS__ )
 #      include <sys/dir.h>
 #    endif
 #  endif /* ?SYSV */
@@ -180,7 +180,10 @@ struct dirent *readdir(dirp)
 } /* end function readdir() */

 #endif /* NO_DIR */
-
+#ifdef __MVS__
+#include <sys/types.h>
+#include <dirent.h>
+#endif

 /**********************/
 /* Function do_wild() */   /* for porting: dir separator; match(ignore_case) */
